server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -698,53 +698,55 @@ private synchronized IndexService createIndexService(
         List<IndexEventListener> builtInListeners,
         IndexingOperationListener... indexingOperationListeners
     ) throws IOException {
-        final IndexSettings idxSettings = new IndexSettings(indexMetadata, settings, indexScopedSettings);
-        // we ignore private settings since they are not registered settings
-        indexScopedSettings.validate(indexMetadata.getSettings(), true, true, true);
-        logger.debug(
-            "creating Index [{}], shards [{}]/[{}] - reason [{}]",
-            indexMetadata.getIndex(),
-            idxSettings.getNumberOfShards(),
-            idxSettings.getNumberOfReplicas(),
-            indexCreationContext
-        );
+        try (var ignored = threadPool.getThreadContext().stashContext()) {
+            final IndexSettings idxSettings = new IndexSettings(indexMetadata, settings, indexScopedSettings);
+            // we ignore private settings since they are not registered settings
+            indexScopedSettings.validate(indexMetadata.getSettings(), true, true, true);
+            logger.debug(
+                "creating Index [{}], shards [{}]/[{}] - reason [{}]",
+                indexMetadata.getIndex(),
+                idxSettings.getNumberOfShards(),
+                idxSettings.getNumberOfReplicas(),
+                indexCreationContext
+            );
 
-        final IndexModule indexModule = new IndexModule(
-            idxSettings,
-            analysisRegistry,
-            getEngineFactory(idxSettings),
-            directoryFactories,
-            () -> allowExpensiveQueries,
-            indexNameExpressionResolver,
-            recoveryStateFactories
-        );
-        for (IndexingOperationListener operationListener : indexingOperationListeners) {
-            indexModule.addIndexOperationListener(operationListener);
-        }
-        pluginsService.forEach(p -> p.onIndexModule(indexModule));
-        for (IndexEventListener listener : builtInListeners) {
-            indexModule.addIndexEventListener(listener);
+            final IndexModule indexModule = new IndexModule(
+                idxSettings,
+                analysisRegistry,
+                getEngineFactory(idxSettings),
+                directoryFactories,
+                () -> allowExpensiveQueries,
+                indexNameExpressionResolver,
+                recoveryStateFactories
+            );
```

Comment:
As well as here:
```
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.IndexModule.IndexModule.<init>>(IndicesService.java:713)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.indices.IndicesService.createIndexService(IndicesService.java:713)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:622)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:175)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.indices.cluster.IndicesClusterStateService.createIndicesAndUpdateShards(IndicesClusterStateService.java:476)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:226)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:538)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:524)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:497)
```

Reply:
In both cases deprecation headers remain in thread context until `callClusterStateListeners` is executed that might contain listeners for unrelated changes (such as other index create or update operations).

Similar issue exists in 7.17 branch: https://github.com/elastic/elasticsearch/pull/93190

Warning should be whitelisted for all 7.17 versions that are already released.